### PR TITLE
Problem: missing c++ bindings for dynamic contract call, query (fix #…

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution: add dynamic contract encode , call

### CODE REVIEW

**Comment by leejw51crypto** [comment]
rebasing now

File: bindings/cpp/src/ethereum.rs [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    pub fn call(
+        &mut self,
+        rpcserver: &str,
+        contract_address: &str,
+        function_name: &str,
+    ) -> Result<String> {
+        let rt = tokio::runtime::Runtime::new().map_err(|_err| EthError::AsyncRuntimeError)?;
+        let res = rt.block_on(self.do_call(rpcserver, contract_address, function_name))?;
+        Ok(res)
+    }
```
Comment (tomtau) [inline bindings/cpp/src/ethereum.rs:305]: you can use `DynamicContract` as mentioned on Slack:
1. `let dc = DynamicContract::new(…)`
2. `let contract_call = dc.function_call(…)`

**Review by tomtau — CHANGES_REQUESTED** [review]
`DynamicContract` should be used -- it will save a lot of boilerplate (it will do all the encoding, decoding, and validation): https://github.com/crypto-com/defi-wallet-core-rs/blob/main/common/src/contract.rs#LL96

`DynamicContract` works with the same wrappers as ERC20, ERC721 and ERC1155 contracts (you can construct it with `Contract::new_dynamic`)

after calling `functional_call`, you'll get `ContractCall` (as with ERC20, ERC721 and ERC1155), so you use the same methods as with those fixed contracts:

* `get_tx` if you want the raw tx
* `estimate_gas` if you want to estimate its gas cost
* `send` if it's a transaction
* `call` if it's a blockchain query

File: bindings/cpp/src/ethereum.rs (hunk 1 above)
Comment (leejw51crypto) [reply bindings/cpp/src/ethereum.rs:305]: ok, i'll modify to use DynamicContract

File: example/cpp-example/cronos.cc [hunk 2]
```diff
@@ -493,6 +521,111 @@ void test_erc20_total_supply() {
...
+  mycontractcall->add_uint("3000");
+  Box<EthAbiTokenWrapper> datafixedarray = mycontractcall->commit_fixed_array();
+
+  mycontractcall->begin_tuple();
+  mycontractcall->add_uint("1");
+  mycontractcall->add_wrapper(datatuple);
+  mycontractcall->add_wrapper(dataarray);
+  mycontractcall->add_wrapper(datafixedarray);
+  mycontractcall->add_uint("2");
+  Box<EthAbiTokenWrapper> finaltuple = mycontractcall->commit_tuple();
```
Comment (tomtau) [inline example/cpp-example/cronos.cc:624]: I find it a little bit hard to follow what the end result type is expected here, so it may be good to write a one-line comment in the Solidity notation to show what's being constructed

**Review by tomtau — CHANGES_REQUESTED** [review]
the low-level data construction API is perhaps ok, but may be a bit hard to use... in comparison with ChainSafe Unity SDK, which just has a single function that expects a String... those low-level API may still be useful for some high-level C++ wrapper (like a template function that one can just pass a vector of UInts, and the C++ template will be translated into that: `begin_...`, loop of `add_unit` and `commit_...`).

so that low-level API can be kept, but for the starter, it may also be good to include a simple one that just takes a `String` or `&str` and parses it into ABI tokens from JSON.

File: example/cpp-example/cronos.cc (hunk 2 above)
Comment (leejw51crypto) [reply example/cpp-example/cronos.cc:624]: ok, i'll add more comment

File: example/cpp-example/cronos.cc [hunk 3]
```diff
@@ -493,6 +521,111 @@ void test_erc20_total_supply() {
...
+  eth_tx_info.data = data;
+  eth_tx_info.gas_limit = "219400";
+  eth_tx_info.gas_price = "100000000";
+  eth_tx_info.gas_price_unit = EthAmount::WeiDecimal;
+
+  Vec<uint8_t> signedtx =
+      build_eth_signed_tx(eth_tx_info, chainid, true, *privatekey);
+  String status =
+      broadcast_eth_signed_raw_tx(signedtx, mycronosrpc, 1000).status;
+  cout << "status: " << status << endl;
```
Comment (tomtau) [inline example/cpp-example/cronos.cc:570]: with `DynamicContract` + String API, it can be something like:
```cpp
DynamicContractWrapper contract = new_dynamic_contract(mycontract, json,
                             mycronosrpc, chainid);
std::stringstream buffer;
buffer << "[" << senderAddress << ", " << receiverAddress << ", " << tokenid << "]";
std::string params = buffer.str();
String status = contract
               .send("safeTransferFrom", params,
                                   *privatekey)
               .status;
```

Comment (leejw51crypto) [reply example/cpp-example/cronos.cc:570]: ok, i'll check

Comment (tomtau) [reply example/cpp-example/cronos.cc:570]: you may need to create some extra helper similar to `broadcast_contract_transfer_tx_blocking` etc. or @damoncro can advise whether any of the existing contract broadcast/query functions could be extended / refactored to allow for dynamic contracts

Comment (damoncro) [reply example/cpp-example/cronos.cc:570]: Let me check~

**Comment by leejw51crypto** [comment]
ok, i'll add minting abi-json, run as a sample,
currently, it's erc721

**Comment by leejw51crypto** [comment]
currently working with DynamicContract and Detokenize to return serde json string,
after that, will fix CI errors

**Comment by leejw51crypto** [comment]
making simpler

File: bindings/cpp/src/ethereum.rs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            rpcserver: &str,
+            contract_address: &str,
+            function_name: &str,
+            function_args: &str, // json
+        ) -> Result<String>;
+
+    }
+} // end of ffi
+
+fn new_eth_contract(abi_json: String) -> Result<Box<EthContract>> {
```
Comment (tomtau) [inline bindings/cpp/src/ethereum.rs:59]: should this take `contract_address` and `rpcserver`, so that `Provider` and `DynamicContract` can be constructed just once?

File: example/cpp-example/cronos.cc [hunk 5]
```diff
@@ -493,6 +504,92 @@ void test_erc20_total_supply() {
...
+  eth_tx_info.amount_unit = EthAmount::EthDecimal;
+  eth_tx_info.data = data;
+  eth_tx_info.gas_limit = "219400";
+  eth_tx_info.gas_price = "100000000";
+  eth_tx_info.gas_price_unit = EthAmount::WeiDecimal;
+
+  Vec<uint8_t> signedtx =
+      build_eth_signed_tx(eth_tx_info, chainid, true, *privatekey);
+  CronosTransactionReceiptRaw receipt =
+      broadcast_eth_signed_raw_tx(signedtx, mycronosrpc, 1000);
```
Comment (tomtau) [inline example/cpp-example/cronos.cc:559]: it's still helpful to have it split like written here, so that one can send the unsigned encoded tx for walletconnect or other signers.

but I was wondering if `EthContract` could have a `send` method that will do all the steps (signing and broadcast), similar to https://github.com/crypto-com/defi-wallet-core-rs/blob/main/bindings/cpp/src/contract.rs#LL67 ?

**Review by tomtau — COMMENTED** [review]
looks simpler! just a small nit that EthContract could take `rpcserver: &str` and `contract_address: &str` in the constructor, so that these params don't need to be repeated in other methods

File: bindings/cpp/src/ethereum.rs (hunk 4 above)
Comment (leejw51crypto) [reply bindings/cpp/src/ethereum.rs:59]: ok

**Comment by leejw51crypto** [comment]
modifying and will add erc721 minting example

File: Cargo.lock [hunk 6]
```diff
@@ -5062,43 +5245,100 @@ version = "0.36.1"
...
- "windows_aarch64_msvc",
- "windows_i686_gnu",
- "windows_i686_msvc",
- "windows_x86_64_gnu",
- "windows_x86_64_msvc",
+ "windows_aarch64_msvc 0.36.1",
+ "windows_i686_gnu 0.36.1",
+ "windows_i686_msvc 0.36.1",
+ "windows_x86_64_gnu 0.36.1",
+ "windows_x86_64_msvc 0.36.1",
```
Comment (damoncro) [inline Cargo.lock:5252]: May I ask why and how bringing `abi-contract` into Cargo.toml will make a big update on Cargo.lock? Why we use the 0.36.1 version here?

Comment (damoncro) [reply Cargo.lock:5252]: Just curious...

**Comment by damoncro** [comment]
NO issues, LGTM.

**Review by damoncro — APPROVED** [review]

File: bindings/cpp/src/ethereum.rs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    abi_json: String,
+) -> Result<Box<EthContract>> {
+    Ok(Box::new(EthContract {
+        rpcserver,
+        contract_address,
+        abi_json,
+    }))
+}
+
+impl EthContract {
```
Comment (damoncro) [inline bindings/cpp/src/ethereum.rs:73]: Missing `do_send` and `send` functions.

File: example/cpp-example/cronos.cc (hunk 3 above)
Comment (damoncro) [reply example/cpp-example/cronos.cc]: Refactoring the existing ERC20/ERC721/ERC1155 functions are possible. It could make all functions united and ruled under Dynamic Contracts. But I guess no big benefits, as those functions are standard functions.

But refactoring is still good because it could extending the missing extension functions, which we are missing and untested, for example enumeration on ERC721.

File: bindings/cpp/src/ethereum.rs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+use ethers::abi::InvalidOutputType;
+use ethers::abi::Token;
+use ethers::prelude::*;
+use ethers::types::transaction::eip2718::TypedTransaction;
+use std::convert::TryFrom;
+
+pub struct EthContract {
+    abi_json: String,
+    rpcserver: String,
+    contract_address: String,
```
Comment (tomtau) [inline bindings/cpp/src/ethereum.rs:16]: instead of these three fields, it could just have `contract: DynamicContract` that will be created in `new_eth_contract`, so that `encode` and `call` can just do `self.contract.(...)` instead of creating the contract over and over?

File: bindings/cpp/src/ethereum.rs (hunk 7 above)
Comment (leejw51crypto) [reply bindings/cpp/src/ethereum.rs:88]: ok

File: bindings/cpp/src/ethereum.rs (hunk 8 above)
Comment (leejw51crypto) [reply bindings/cpp/src/ethereum.rs:19]: ok, i'll check

File: Cargo.lock (hunk 6 above)
Comment (leejw51crypto) [reply Cargo.lock:5252]: to use DynamicContract

File: common/src/transaction/ethereum/abi_contract.rs [hunk 9]
```diff
@@ -3,9 +3,10 @@
 use crate::node::ethereum::abi::EthAbiToken;
 use crate::EthError;
 use ethers::prelude::abi::{Contract, Token};
-
+use serde::{Deserialize, Serialize};
 /// Ethereum ABI token to ffi bind
-#[derive(Debug, Eq, PartialEq)]
+// enable serde serialization
+#[derive(Debug, Eq, PartialEq, Serialize, Deserialize)]
```
Comment (tomtau) [inline common/src/transaction/ethereum/abi_contract.rs:9]: is the default automatic serde derive ok? I just worry about those variants that have `Vec<u8>` that the default one will require a json array of numbers which is a bit inconvenient / wasteful, but it seems it's like that in ethabi/ethers anyway

**Review by tomtau — APPROVED** [review]
